File: silverwind-frontend/src/app/core/services/profile.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse } from '../models/auth.model';

export interface Document {
  id: string;
  documentType: string;
  documentName: string;
  fileUrl: string;
  createdAt: string;
}

export interface Education {
  id: string;
  institution: string;
  degree: string;
  fieldOfStudy?: string;
  startDate?: string;
  endDate?: string;
  grade?: string;
  description?: string;
}

export interface Certification {
  id: string;
  name: string;
  issuingOrganization: string;
  issueDate?: string;
  expirationDate?: string;
  credentialId?: string;
  credentialUrl?: string;
}

export interface WorkExperience {
  id: string;
  jobTitle: string;
  companyName: string;
  location?: string;
  startDate: string;
  endDate?: string;
  currentJob: boolean;
  description?: string;
}

export interface Skill {
  id: string;
  skillName: string;
  proficiencyLevel?: string;
  yearsOfExperience?: number;
}

@Injectable({
  providedIn: 'root',
})
export class ProfileService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/employees`;

  // --- Documents ---
  getDocuments(employeeId: string): Observable<ApiResponse<Document[]>> {
    return this.http.get<ApiResponse<Document[]>>(`${this.apiUrl}/${employeeId}/documents`);
  }

  uploadDocument(
    employeeId: string,
    file: File,
    type: string,
    name?: string,
  ): Observable<ApiResponse<Document>> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('type', type);
    if (name) formData.append('name', name);

    return this.http.post<ApiResponse<Document>>(
      `${this.apiUrl}/${employeeId}/documents`,
      formData,
    );
  }

  deleteDocument(employeeId: string, documentId: string): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(
      `${this.apiUrl}/${employeeId}/documents/${documentId}`,
    );
  }

  downloadDocumentUrl(employeeId: string, documentId: string): string {
    return `${this.apiUrl}/${employeeId}/documents/${documentId}/download`;
  }

  downloadDocument(employeeId: string, documentId: string): Observable<Blob> {
    return this.http.get(`${this.apiUrl}/${employeeId}/documents/${documentId}/download`, {
      responseType: 'blob',
    });
  }

  // --- Education ---
  getEducation(employeeId: string): Observable<ApiResponse<Education[]>> {
    return this.http.get<ApiResponse<Education[]>>(`${this.apiUrl}/${employeeId}/education`);
  }

  addEducation(
    employeeId: string,
    education: Partial<Education>,
  ): Observable<ApiResponse<Education>> {
    return this.http.post<ApiResponse<Education>>(
      `${this.apiUrl}/${employeeId}/education`,
      education,
    );
  }

  deleteEducation(employeeId: string, educationId: string): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(
      `${this.apiUrl}/${employeeId}/education/${educationId}`,
    );
  }

  // --- Certifications ---
  getCertifications(employeeId: string): Observable<ApiResponse<Certification[]>> {
    return this.http.get<ApiResponse<Certification[]>>(
      `${this.apiUrl}/${employeeId}/certifications`,
    );
  }

  addCertification(
    employeeId: string,
    certification: Partial<Certification>,
  ): Observable<ApiResponse<Certification>> {
    return this.http.post<ApiResponse<Certification>>(
      `${this.apiUrl}/${employeeId}/certifications`,
      certification,
    );
  }

  deleteCertification(employeeId: string, certificationId: string): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(
      `${this.apiUrl}/${employeeId}/certifications/${certificationId}`,
    );
  }

  // --- Work Experience ---
  getWorkExperience(employeeId: string): Observable<ApiResponse<WorkExperience[]>> {
    return this.http.get<ApiResponse<WorkExperience[]>>(
      `${this.apiUrl}/${employeeId}/work-experience`,
    );
  }

  addWorkExperience(
    employeeId: string,
    experience: Partial<WorkExperience>,
  ): Observable<ApiResponse<WorkExperience>> {
    return this.http.post<ApiResponse<WorkExperience>>(
      `${this.apiUrl}/${employeeId}/work-experience`,
      experience,
    );
  }

  deleteWorkExperience(employeeId: string, experienceId: string): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(
      `${this.apiUrl}/${employeeId}/work-experience/${experienceId}`,
    );
  }

  // --- Skills ---
  getSkills(employeeId: string): Observable<ApiResponse<Skill[]>> {
    return this.http.get<ApiResponse<Skill[]>>(`${this.apiUrl}/${employeeId}/skills`);
  }

  addSkill(employeeId: string, skill: Partial<Skill>): Observable<ApiResponse<Skill>> {
    return this.http.post<ApiResponse<Skill>>(`${this.apiUrl}/${employeeId}/skills`, skill);
  }

  deleteSkill(employeeId: string, skillId: string): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${employeeId}/skills/${skillId}`);
  }
}

File: silverwind-backend/src/main/java/com/solventek/silverwind/feature/profile/service/ProfileService.java
package com.solventek.silverwind.feature.profile.service;

import com.solventek.silverwind.auth.Employee;
import com.solventek.silverwind.auth.EmployeeRepository;
import com.solventek.silverwind.feature.profile.entity.*;
import com.solventek.silverwind.feature.profile.repository.*;
import com.solventek.silverwind.storage.StorageService;
import lombok.RequiredArgsConstructor;
import org.springframework.core.io.Resource;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Transactional
public class ProfileService {

    private final EmployeeRepository employeeRepository;
    private final DocumentRepository documentRepository;
    private final EducationRepository educationRepository;
    private final CertificationRepository certificationRepository;
    private final WorkExperienceRepository workExperienceRepository;
    private final SkillRepository skillRepository;
    private final StorageService storageService;

    // --- Helper to get Employee ---
    private Employee getEmployee(UUID employeeId) {
        return employeeRepository.findById(employeeId)
                .orElseThrow(() -> new RuntimeException("Employee not found with ID: " + employeeId));
    }

    // --- Documents ---
    public List<Document> getDocuments(UUID employeeId) {
        return documentRepository.findByEmployeeId(employeeId);
    }

    public Document uploadDocument(UUID employeeId, MultipartFile file, String documentType, String documentName) {
        Employee employee = getEmployee(employeeId);

        String storageKey = storageService.upload(file, "employees/" + employee.getEmail() + "/documents");
        String fileUrl = storageService.getPresignedUrl(storageKey, null); // Or just null if dynamic

        Document document = Document.builder()
                .employee(employee)
                .documentType(documentType)
                .documentName(documentName != null ? documentName : file.getOriginalFilename())
                .fileUrl(fileUrl) // Store initial URL, but typically regenerate on get
                .storageKey(storageKey)
                .build();

        return documentRepository.save(document);
    }

    public void deleteDocument(UUID documentId) {
        Document document = documentRepository.findById(documentId)
                .orElseThrow(() -> new RuntimeException("Document not found"));
        
        if (document.getStorageKey() != null) {
            try {
                storageService.delete(document.getStorageKey());
            } catch (Exception e) {
                // Log but proceed
            }
        }
        documentRepository.delete(document);
    }

    public Resource downloadDocument(UUID documentId) {
        Document document = documentRepository.findById(documentId)
                .orElseThrow(() -> new RuntimeException("Document not found"));
        return storageService.download(document.getStorageKey());
    }

    // --- Education ---
    public List<Education> getEducation(UUID employeeId) {
        return educationRepository.findByEmployeeIdOrderByStartDateDesc(employeeId);
    }

    public Education addEducation(UUID employeeId, Education education) {
        education.setEmployee(getEmployee(employeeId));
        return educationRepository.save(education);
    }

    public Education updateEducation(UUID educationId, Education educationDetails) {
        Education education = educationRepository.findById(educationId)
                .orElseThrow(() -> new RuntimeException("Education not found"));
        
        education.setInstitution(educationDetails.getInstitution());
        education.setDegree(educationDetails.getDegree());
        education.setFieldOfStudy(educationDetails.getFieldOfStudy());
        education.setStartDate(educationDetails.getStartDate());
        education.setEndDate(educationDetails.getEndDate());
        education.setGrade(educationDetails.getGrade());
        education.setDescription(educationDetails.getDescription());

        return educationRepository.save(education);
    }

    public void deleteEducation(UUID educationId) {
        educationRepository.deleteById(educationId);
    }

    // --- Certifications ---
    public List<Certification> getCertifications(UUID employeeId) {
        return certificationRepository.findByEmployeeIdOrderByIssueDateDesc(employeeId);
    }

    public Certification addCertification(UUID employeeId, Certification certification) {
        certification.setEmployee(getEmployee(employeeId));
        return certificationRepository.save(certification);
    }

    public void deleteCertification(UUID certificationId) {
        certificationRepository.deleteById(certificationId);
    }

    // --- Work Experience ---
    public List<WorkExperience> getWorkExperience(UUID employeeId) {
        return workExperienceRepository.findByEmployeeIdOrderByStartDateDesc(employeeId);
    }

    public WorkExperience addWorkExperience(UUID employeeId, WorkExperience experience) {
        experience.setEmployee(getEmployee(employeeId));
        return workExperienceRepository.save(experience);
    }

    public void deleteWorkExperience(UUID experienceId) {
        workExperienceRepository.deleteById(experienceId);
    }

    // --- Skills ---
    public List<Skill> getSkills(UUID employeeId) {
        return skillRepository.findByEmployeeId(employeeId);
    }

    public Skill addSkill(UUID employeeId, Skill skill) {
        skill.setEmployee(getEmployee(employeeId));
        return skillRepository.save(skill);
    }

    public void deleteSkill(UUID skillId) {
        skillRepository.deleteById(skillId);
    }
}
